// _egg-ts-helper@1.22.3@egg-ts-helper/dist/scripts/plugin.js
"use strict";
/**
 * Getting plugin info in child_process to prevent effecting egg application( splitting scopes ).
 */
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const fs_1 = tslib_1.__importDefault(require("fs"));
const path_1 = tslib_1.__importDefault(require("path"));
const utils_1 = require("../utils");
const url = process.argv[2];
const stat = fs_1.default.statSync(url);
let pluginInfos = {};
if (stat.isDirectory()) {
    const framework = (utils_1.getPkgInfo(url).egg || {}).framework || 'egg';
    const loader = getLoader(url, framework);
    if (loader) {
        loader.loadPlugin();
        loader.loadConfig();
        pluginInfos = loader.allPlugins;
    }
}
else if (stat.isFile()) {
    const config = utils_1.requireFile(url);
    for (const name in config) {
        const plugin = config[name];
        if (typeof plugin === 'boolean') {
            pluginInfos[name] = { enable: plugin };
        }
        else {
            pluginInfos[name] = {};
            ['package', 'enable'].forEach(k => {
                if (k in plugin) {
                    pluginInfos[name][k] = plugin[k];
                }
            });
        }
    }
}
process.stdout.write(JSON.stringify(pluginInfos));
function noop() { }
function getLoader(baseDir, framework) {
    const frameworkPath = path_1.default.join(baseDir, 'node_modules', framework);
    const eggCore = findEggCore(baseDir, frameworkPath);
    if (!eggCore)
        return;
    const EggLoader = eggCore.EggLoader;
    const egg = utils_1.requireFile(frameworkPath);
    if (!egg)
        return;
    process.env.EGG_SERVER_ENV = 'local';
    return new EggLoader({
        baseDir,
        logger: {
            debug: noop,
            info: noop,
            warn: noop,
            error: noop,
        },
        app: Object.create(egg.Application.prototype),
    });
}
function findEggCore(baseDir, frameworkPath) {
    let eggCorePath = path_1.default.join(baseDir, 'node_modules/egg-core');
    if (!fs_1.default.existsSync(eggCorePath)) {
        eggCorePath = path_1.default.join(frameworkPath, 'node_modules/egg-core');
    }
    // try to load egg-core in cwd
    const eggCore = utils_1.requireFile(eggCorePath);
    if (!eggCore) {
        // try to resolve egg-core
        return utils_1.resolveModule('egg-core');
    }
    return eggCore;
}
